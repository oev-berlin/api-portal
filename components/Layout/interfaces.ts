import React from 'react';

export interface LayoutProps {
    searchOptions: string[],
    children: React.ReactNode,
}
