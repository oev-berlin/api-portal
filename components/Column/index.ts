import { Column } from './Column';
import { ColumnProps } from './interfaces';

export { Column };
export type { ColumnProps };
