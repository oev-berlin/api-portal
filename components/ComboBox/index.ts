import ComboBox from './ComboBox';
import { ComboBoxProps } from './interfaces';

export { ComboBox };
export type { ComboBoxProps };
