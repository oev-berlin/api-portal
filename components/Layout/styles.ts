import styled from 'styled-components';
import { Toolbar } from '@mui/material';

export const Title = styled.h1`
    font-weight: bold;
`;

export const StyledToolBar = styled(Toolbar)`
  position: static;
  display: flex;
  flex-wrap: wrap;
  padding: 1rem;
  flex-grow: 1;
  justify-content: space-around;
`;
