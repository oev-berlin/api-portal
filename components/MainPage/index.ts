export {CustomisedCard} from "./ProjectCard";
export { MainPage } from './MainPage';
