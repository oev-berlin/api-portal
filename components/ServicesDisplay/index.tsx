import { ServicesDisplay } from './ServicesDisplay';

export { ServicesDisplay };
