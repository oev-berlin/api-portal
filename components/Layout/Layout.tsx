import React from 'react';
import { ComboBox } from '../ComboBox';
import { LayoutProps } from './interfaces';
import { StyledToolBar, Title } from './styles';

export const Layout = ({ searchOptions, children }: LayoutProps) => (
  <>
    <StyledToolBar>
      <Title>Swagger API</Title>
      <ComboBox searchOptions={searchOptions} />
    </StyledToolBar>
    {children}
  </>
);
