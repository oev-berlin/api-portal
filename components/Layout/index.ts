import { Layout } from './Layout';
import { LayoutProps } from './interfaces';

export { Layout };
export type { LayoutProps };
