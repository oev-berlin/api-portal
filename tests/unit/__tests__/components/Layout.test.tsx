import { act, render, cleanup, waitFor } from '@testing-library/react';
import { axe, toHaveNoViolations} from 'jest-axe';
import React from 'react';
import { Layout, LayoutProps } from '../../../../components/Layout';
import 'jest-styled-components';
import userEvent from '@testing-library/user-event';

expect.extend(toHaveNoViolations);

export const useRouter = jest.spyOn(require("next/router"), "useRouter");
const pushMock = jest.fn()
const setupComponent = ({ searchOptions, children }: LayoutProps) => render(
  <Layout
    searchOptions={searchOptions}
    children={children}
  />,
);

describe('Layout', () => {
  afterEach(() => {
    cleanup();
  });

  describe('UI Tests', () => {
    useRouter.mockImplementationOnce(()=>({
      pathname: "/",
      route: "/",
      query: "",
      asPath: "/",})
    );
    it('Layout component should render without crashing', async () => {
      const screen = setupComponent({
        searchOptions: ['salneo', 'tidy'],
        children: <div />,
      });
      expect(screen.container).toBeDefined();
    });
    it('Layout component (searchOptions is empty) should render without crashing', async () => {
      const screen = setupComponent({
        searchOptions: [""],
        children: <div />,
      });
      expect(screen.container).toBeDefined();
    });

    it('Layout component should render correctly (Title should be rendered)', async () => {
      const screen = setupComponent({ searchOptions: ['salneo', 'tidy'], children: <div /> });
      await screen
        .findByText(/swagger api/i)
        .then((element) => expect(element).toBeInTheDocument());
    });
    it('Layout component should render correctly (Title should be rendered even if options are empty)', async () => {
      const screen = setupComponent({ searchOptions: [], children: <div /> });
      await screen
          .findByText(/swagger api/i)
          .then((element) => expect(element).toBeInTheDocument());
    });

    it('Layout component should behave correctly (click on the arrow)', async () => {
      const screen =  setupComponent({ searchOptions: ['salneo', 'tidy'], children: <div /> });
      expect(await screen.queryByText(/salneo/)).not.toBeInTheDocument();
      expect(await screen.queryByText(/tidy/)).not.toBeInTheDocument();

      await act(() => screen.findByRole('button')
          .then((button) => userEvent.click(button)));

      await screen.findByText(/salneo/).then((element) => expect(element).toBeInTheDocument());
      await screen.findByText(/tidy/).then((element) => expect(element).toBeInTheDocument());

    });
    it('Layout component should behave correctly (click on an item from the list)', async () => {
      useRouter.mockReturnValue({
        query: "/",
        push: pushMock
      })
      const screen =  setupComponent({ searchOptions: ['salneo', 'tidy'], children: <div /> });
      await act(() => screen.findByRole('button')
          .then((button) => userEvent.click(button)));

      await act(() =>screen.findByText(/salneo/)
          .then((salneoOption) => userEvent.click(salneoOption)));
      expect(pushMock).toHaveBeenCalledWith("/project/salneo");
    });
    it('Layout component should behave correctly (Autocomplete should work correctly)', async () => {
      useRouter.mockReturnValue({
        query: "/",
        push: pushMock
      })
      const screen =  setupComponent({ searchOptions: ['salneo', 'tidy'], children: <div /> });
      expect(await screen.queryByText(/salneo/)).not.toBeInTheDocument();
      expect(await screen.queryByText(/tidy/)).not.toBeInTheDocument();

      await act(() => screen.findByRole('combobox')
          .then((combobox) => userEvent.type(combobox, 's')));

      expect(await screen.queryByText(/salneo/)).toBeInTheDocument();
      expect(await screen.queryByText(/tidy/)).not.toBeInTheDocument();
      await act(() =>screen.findByText(/salneo/)
          .then((salneoOption) => userEvent.type(salneoOption, '{enter}')));
      await act(() => screen.findByRole('combobox').then((combobox)=>expect(combobox.value).toEqual('salneo')))

    });
  });

  describe('Snapshots Tests', () => {
    it('should match a basic snapshot', () => {
      const { container } = setupComponent({ searchOptions: ['salneo', 'tidy'], children: <div /> });
      expect(container).toMatchSnapshot();
    });
    it('should match a basic snapshot (searchOptions is empty)', () => {
      const { container } = setupComponent({ searchOptions: [""], children: <div /> });
      expect(container).toMatchSnapshot();
    });
  });
  describe('Accessibility Tests', () => {
    it('should pass a basic accessibility test', async () => {
      const { container } = setupComponent({ searchOptions: ['salneo'], children: <div /> });
      const results = await waitFor(() => axe(container));
      expect(results).toHaveNoViolations();
    });
  });
});