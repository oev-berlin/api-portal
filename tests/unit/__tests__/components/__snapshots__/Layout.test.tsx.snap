// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Snapshots Tests should match a basic snapshot (searchOptions is empty) 1`] = `
.c2 {
  width: 300px;
}

.c1 {
  font-weight: bold;
}

.c0 {
  position: static;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 1rem;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

<div>
  <div
    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular c0 css-hyum1k-MuiToolbar-root"
  >
    <h1
      class="c1"
    >
      Swagger API
    </h1>
    <div
      class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon c2 css-1qqsdnr-MuiAutocomplete-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="mui-10"
          id="mui-10-label"
        >
          Project
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-10"
            role="combobox"
            spellcheck="false"
            type="text"
            value=""
          />
          <div
            class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Open"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
              tabindex="-1"
              title="Open"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Project
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Layout Snapshots Tests should match a basic snapshot 1`] = `
.c2 {
  width: 300px;
}

.c1 {
  font-weight: bold;
}

.c0 {
  position: static;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 1rem;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

<div>
  <div
    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular c0 css-hyum1k-MuiToolbar-root"
  >
    <h1
      class="c1"
    >
      Swagger API
    </h1>
    <div
      class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon c2 css-1qqsdnr-MuiAutocomplete-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="mui-8"
          id="mui-8-label"
        >
          Project
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-8"
            role="combobox"
            spellcheck="false"
            type="text"
            value=""
          />
          <div
            class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Open"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
              tabindex="-1"
              title="Open"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Project
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;
